### ISSUE TITLE
[Brainstorming] MCore Module System

### ISSUE THREAD

**Issue by johnwikman** [miking-lang/miking#125]
Now that the standard library is starting to grow, it is high time to start talking about the MCore module system. This brainstorming issue serves as a general discussion hub for the MCore module system where people can share links, suggest features, and bring forward opinions.

The current module system looks like this (as of 2020-08-04):
```
-- file: fib.mc
let fib = lam n.
  recursive let work = lam i. lam prev. lam curr.
    if eqi i n then
      curr
    else
      work (addi i 1) curr (addi prev curr)
  in
  work 0 1 0

---------------------------
-- file: foo.mc

-- "copy-paste" the contents of the fib.mc file here
include "fib.mc"

let foo = lam n. muli (fib n) 2

utest foo 0 with 0
utest foo 1 with 2
utest foo 3 with 4
```

# Brainstorming Summary
This is a summary of what has been suggested or shared in this issue so far. This summary should be updated as the discussion progresses.

## Shared Links
- [Elm Module System](https://guide.elm-lang.org/webapps/modules.html#using-modules)

## Suggested Features
- [Protect identifiers from being overwritten by module definitions](https://github.com/miking-lang/miking/issues/125#issuecomment-668652457)

**Comment by johnwikman** [comment]
A feature that I would like to have embedded into the module system is to be able to protect certain identifiers such that they do not get overwritten by a definition from the included module. For example, if we need to include a library that overwrites the intrinsic function `eqf`:
```
-- file: superfloat.mc
let tolerance = 1.0

-- Redefine float equality to be within our tolerance
let eqf: Float -> Float -> Bool = lam x. lam y.
  let diff = if ltf x y then subf y x else subf x y in
  leqf diff tolerance

...
(lots of important functions defined here, which depend on the definition of eqf above)
let factorialf = Float -> Float = ...

------------------------------------------------

-- file: foo.mc
include "superfloat.mc"

-- This will fail since superfloat.mc redefined eqf, but I still need the other functions defined in superfloat.mc!
utest eqf 1.0 1.5 with false

let foo = lam x. addf (factorialf x) x
```

While it could be argued that the issue here is that the library is not well implemented, this example is very simple. When this problem appears in a more complex situation it would be good to have a way to solve this. Here a couple of ways which this can be resolved using concepts from other languages:

#### Namespaces (C++)
```
namespace sf
  include "superfloat.mc"
endns

utest eqf 1.0 1.5 with false -- this will no longer fail

let foo = lam x. addf (sf::factorialf x) x
```

#### Specific Identifier Imports (Python)
```
import factorialf otherf from "superfloat.mc"

utest eqf 1.0 1.5 with false -- this will also no longer fail since eqf is not imported

let foo = lam x. addf (factorialf x) x
```

These 2 examples are just existing ways of how this issue can be solved in other languages. There are probably better ways to solve this for MCore.

**Comment by dlunde** [comment]
I like the module systems of both Haskell and OCaml. In general, I do not have any strong opinions on this matter.

**Comment by elegios** [comment]
I rather like how things work in Haskell (surprise, surprise), but I would strongly suggest changing the defaults. In Haskell it's easy to import all names from a module into the current module, but this turns out to rarely be what you want. Elm has swapped this around and is probably a better model to follow if that's the direction we're going in (see https://guide.elm-lang.org/webapps/modules.html#using-modules).

**Comment by johnwikman** [comment]
I like the import syntax for Elm, clean and simple (and familiar to those who have used Python). The question in this case would IMO more be regarding the `exposing` keyword. To me that could be solved by two let statements after the import:
```
import Alice as Bob exposing (foo, bar)

-- to me this is just syntactic sugar for:
import Alice as Bob
let foo = Bob.foo
let bar = Bob.bar
```

I feel like the `exposing` keyword would be complicating things too much in the IR with little gain.

I am also not a big fan of the module manifests that have seemed to pop up in recent languages. In the case of Elm, I would much rather specify which individual labels that are local to this file and not to be exported:
```
-- name.mc
--
-- I do _not_ want to have to write this:
--
-- module Name exposing (nameNoSym)
--
-- This means that I have to essentially maintain a definition in 2 places. If I change the name
-- of the function, I have to change it here as well as where it is defined.

-- noSymbol will not be available to those who include this module
local let noSymbol = gensym ()

let nameNoSym : String -> Name =
  lam x. (x, noSymbol)
```
This keeps the property as close to the function definition as possible and the overall boilerplate on the file is reduced. Though I am open to suggestions on this front.

**Comment by johnwikman** [comment]
Another thing that has been on my mind (and I might be mentally scarred from C) is to include the filesystem location of the module in the import statement. If I write `import SQL`, how do I know where that is coming from? Is it from a specific centralized module repository or locally imported?

An idea could be to expose exactly where that module is located in the filesystem. For example:
```haskell
-- The identifiers Name, Seq, Matrix, and Sqlite3 are not present in the specified
-- .mc files, but instead introduced in this file as an alias to the library. Could
-- as well write   import SQL by <sqlite3.mc>   if we want.
import Name by "name.mc"
import Seq by "seq.mc"
import Matrix by <linalg/matrix.mc>
import Sqlite3 by <sqlite3.mc>

let conn = Sqlite3.connect("localhost:4321")
-- The file sqlite3.mc only defines the function as "let connect = lam addr. ..."
```
Here, an import statement with `"quotes"` strictly implies that the path is looked up relative to the directory of the source file. An import statement with `<angle brackets>` strictly implies that the path is looked up relative to a library folder, either specified as a compiler argument or part of a `$MCORE_LIB_PATH` environment variable.

This is not super elegant, but it does not hide imports behind layers of magic which aids traceability imo. The statement `import Matrix by <linalg/matrix.mc>` is also more effort to write than just `import Matrix`, but for an IR this is probably ok. The naming of `import ALIAS by "file/path.mc"` is also not very intuitive. I tried some other alternatives, but I feel like they had some other issues instead:
```haskell
-- Here is looks like the file "seq.mc" actually contains something called Seq, when we
-- are instead just creating an alias for the file itself.
import Seq from "seq.mc"

-- This is much more intuitive, but it does not look as nice to me due to that the
-- created aliases are misaligned.
import "name.mc" as Name
import "seq.mc" as Seq
import <linalg/matrix.mc> as Matrix
import <sqlite3.mc> as Sqlite3
```

**Comment by elegios** [comment]
That feels a decent amount like the imports of go or javascript (`import "math"` or `import m "math"`, and `import * as Math from "math"`, respectively). I like being explicit on whether imports come from local files or a library, but I'm not entirely sure `""` vs `<>` is the way to go there, it would work just as well to prefix with `./` for local things, and otherwise it's from a library.

It's also worth mentioning that packages (or equivalent, @david-broman has some ideas here I believe) should probably go into the decision here. It's sometimes inconvenient that packages add to the same global namespace in Haskell, it seems nicer to do it in the style of Rust, where a package is a known entity to the language, meaning that the introduced names are introduced in a namespace for each package. 

---------------------------------

Regarding IR, right now modules and includes and stuff only exist on the mlang level, not mexpr, and I wonder if that might be a good thing to retain going forward. Imports and stuff are then only handled on the mlang level, then desugared to a long chain of `let`s with appropriate symbol connections between definitions and uses.

Compiler backends would then presumably pick up each definition along the spine and compile to a separate function/value, so it wouldn't necessarily become just one large function with many local definitions.

The remaining question is then of course how to handle modules and stuff in languages higher up. Do they desugar to mlang, or do they redo the work of connecting names across modules for their own desired semantics?

**Comment by johnwikman** [comment]
I agree with that modules and includes should be on the mlang level. Keeps mexpr as simple as necessary.

Regarding packages; I think that it is good to provide a system of creating versioned packages, keeping track of dependencies, etc. But I'm not so sure that it should be directly integrated into mlang. I am more thinking along the lines that the mlang module system should provide features that enable these kinds of package systems, but not integrate this system straight into the main language/compiler.

**Comment by johnwikman** [comment]
You might have already discussed this internally, but it would be nice to have a module system such that packages can be written in DSLs or other high-level languages that target MCore, and then import that package into a completely separate DSL that also targets MCore. I am not sure if there is a nice way that makes this possible, but it would certainly make module development more pleasant as you are not developing all your code in MCore, as well as reduce code duplication between DSLs.
